### DISCUSSION TITLE
The Github download blues

### DISCUSSION THREAD

**Discussion by krismc** [micropython/micropython#16317] [Libraries & Drivers] [answered]
I'm trying to setup a HX710 driver on an esp32. I downloaded the files from https://github.com/robert-hh/hx710 & put them in a 'hx710' folder on the esp32 then tried to run the first line of a sample program.
```
>>> from hx710 import HX710
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ImportError: no module named 'hx710.HX710'
```
So I'm thinking there is something about how github works that I don't understand?  Is there some install process required rather than simply copying the files onto the esp32?

**Comment by davefes** [comment, marked as answer]
try:
`from hx710_gpio import HX170`
or
`from hx710_pio import HX170`
or
`from hx710_spi import HX170`

**Comment by robert-hh** [comment]
You could do it in different ways. Importing is always done from a directory contained in the `sys.path` list. So if you put you files into a hx710 directory, this directory must be added to sys.path. Usually, files like hx710 would be put into the lib directory, which is part of sys.path be default. So you have at least 3 options:

- add hx710 as the name o the directory to sys.path with `import sys;sys.path.append("/hx710")`. Then you can use e.g. `from hx710_gpio import HX710`. But: you cannot use `from hx710 import HX710`.
- move the file hx710_gpio.py (or any other of the hx710_xxx.py files) to the lib directory. Then you can use e.g. `from hx710_gpio import HX710`.
- keep the hx710 directory, but rename the hx710_xxx.py driver of your choice to `__init__.py`. Note that the hx710 directory itself must be in a place listed by sys.path. Then you can use e.g. `from hx710 import HX710`. hx710 is now the name of the directory.

All of that is standard Python behavior and not related to any specific module.

Note: Please prefer hx710_gpio.py or hx710_pio.py, as the return more precise values.

**Reply by krismc** [reply]
Your https://forum.micropython.org/viewtopic.php?t=5931 post said the spi version was least likely to give a dodgy result Rob?

**Reply by robert-hh** [reply]
That was my 2019 view. Recently, I discovered that actually the SPI version may deliver spurious completely wrong values. The reason is that the HX710/711 issues a data ready pulse at the data line when a conversion is finished. This will happen even when a data transfer is active. In the last version, the -gpio and -pio versions of the driver synchronise the data transfer with the data ready pulse. The -spi version does not. So it will happen with that version that the data ready pulse interferes with the data, leading to a wrong result.

**Reply by krismc** [reply]
I put your files in lib & got the ball rolling with 
```
from hx710_gpio import HX710 
```
I started working through the pycom example on an old gpy I had laying around but didn't get far
```
>>> from machine import Pin
>>> pin_OUT = Pin("P9", Pin.IN, pull=Pin.PULL_DOWN)
>>> pin_SCK = Pin("P10", Pin.OUT)
>>> hx710 = HX710(pin_SCK, pin_OUT)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/flash/lib/hx710_gpio.py", line 39, in __init__
  File "/flash/lib/hx710_gpio.py", line 45, in set_mode
  File "/flash/lib/hx710_gpio.py", line 54, in read
AttributeError: 'Pin' object has no attribute 'irq'
```
Any idea how I can fix this error?

**Reply by robert-hh** [reply]
Thank you for the note. It seems that the firmware does not support the Pycom version any more. Using Pin.irq was added after Pycom closed it's business and stopped ESP32 support. You can use generic ESP32 firmware with a standard ESP32 board or even a Pycom board. I can change the driver to:
a) still support Pycom firmware and
b) remove mentioning Pycom.

**Reply by robert-hh** [reply]
I have updated the hx710 and hx711 repository to deal with a situation where pin.irq() does not exist. It works with Pycom devices.

**Reply by krismc** [reply]
Thx Rob, I ran your new files on a 2040 & got some results
```
>>> hx710.tare(times=15)
>>> hx710.read()
902741
>>> for i in range(5): print(hx710.get_value()); sleep(1)

106.75
331.5
351.5625
471.8125
567.0625
>>> hx710.read()
902371
```
These are Hx710 readings from a differential pressure sensor in air. Since the pressure sensor is not connected to anything I was expecting  the offset get_value to be steady? Have I missed something?

**Reply by krismc** [reply]
Also I took the liberty of sticking a print stmt in your set_offset fn to display the offset
```
>>> from hx710_pio import HX710; from machine import Pin; from time import sleep; pin_OUT = Pin(12, Pin.IN, pull=Pin.PULL_DOWN); pin_SCK = Pin(13, Pin.OUT); hx710 = HX710(pin_SCK, pin_OUT); hx710.tare()
offset 1484502.0
>>> for i in range(5): raw=hx710.read_average(); oset=hx710.get_value(); print(raw, raw-oset, oset); sleep(1)

1484995.0 1484793.0 202.125
1484876.0 1484700.0 176.875
1484524.0 1484284.0 240.25
1484380.0 1484196.0 184.5
1484445.0 1484295.0 150.25
```
I don't understand why the offset in the middle column isn't constant at the offset=1484502.0 ?

**Reply by robert-hh** [reply]
Because you are not printing the offset, as being set by hx710.tare(). You can access that value as hx710.OFFSET. So if you use `print(raw, hx710.OFFSET, oset)`, the middle column should be constant.
The readings of the sensor are quite noisy. I use for testing a fixed resistor bridge as close as possible to the chip pins and still see substantial noise in the range of +/- 200 absolute reading (about 10**-5 of the full range). The effective resolution of the HX710/711 is about 16 bits. The lower 8 bits are usually noise. The CS1237 is slightly better. By averaging or low-pass filtering you can reduce the noise. But both ADCs are not really precise. For scale applications and slow reading that should be sufficient.

**Reply by krismc** [reply]
The print stmt I put in set_offset returns the same result as hx710.OFFSET. What I don't understand is why raw-oset is not the same as hx710.OFFSET? 
```
>>> from hx710_pio import HX710; from machine import Pin; from time import sleep; pin_OUT = Pin(12, Pin.IN, pull=Pin.PULL_DOWN); pin_SCK = Pin(13, Pin.OUT); hx710 = HX710(pin_SCK, pin_OUT); hx710.tare()
offset 1531351.0
>>> for i in range(5): raw=hx710.read_average(); oset=int(hx710.get_value()); print(raw, hx710.OFFSET, oset, raw-oset); sleep(1)

1532029.0 1531351.0 575 1531454.0
1532006.0 1531351.0 593 1531413.0
1532056.0 1531351.0 498 1531558.0
1532011.0 1531351.0 555 1531456.0
1531589.0 1531351.0 484 1531105.0
```

**Reply by robert-hh** [reply]
> What I don't understand is why raw-oset is not the same as hx710.OFFSET

hx710.OFFSET is the value returned by hx710.tare() and is kept at that value until hx710.tare() is called again. Both  hx710.read_average() and hx710.get_value() perform internally hx710.read() (hx710.get_value() = hx710.read() - hx710.OFFSET). Due to sensor and conversion noise, every call of hx710.read() returns a different value, and so hx710.read_average() and hx710.get_value() change for every call.

**Reply by krismc** [reply]
Yeah I understand read_average & get_value vary each time but shouldn't the difference between them, hx710.OFFSET, be the same 1531351 each time since the tare is done only once at the start?

**Reply by robert-hh** [reply]
> but shouldn't the difference between them, hx710.OFFSET, be the same 1531351 each time since the tare is done only once at the start

Why should it? hx710.OFFSET is the value of a read_average() at the time of calling hx710.tare(). Every call pf read_average() of get_value() will return a slightly different value, and so does tare(). Everytime you call hx710.tare(), hx710.OFFSET will change.

**Reply by krismc** [reply]
Are you saying then that there is a new tare done between each of the 5 readings in the example above?
